lib/Driver/Driver.cpp

```diff
@@ -1414,9 +1414,9 @@ void Driver::buildActions(SmallVectorImpl<const Action *> &TopLevelActions,
       }
       case types::TY_SwiftModuleFile:
       case types::TY_SwiftModuleDocFile:
-        if (OI.ShouldGenerateModule) {
-          // When generating a .swiftmodule, treat .swiftmodule files as
-          // inputs to a MergeModule action.
+        if (OI.ShouldGenerateModule && OI.ShouldTreatModuleAsTopLevelOutput) {
+          // When generating a .swiftmodule as a top-level output, treat
+          // .swiftmodule files as inputs to a MergeModule action.
```

Comment:
Hm. I think I was expecting `OI.ShouldGenerateModule && !OI.ShouldLink`. `ShouldTreatModuleAsTopLevelOutput` is still true when you explicitly say `-emit-module-path`, even if you're also linking.

This also needs some new tests for the mixed-input case (some .swift, some .swiftmodule, with/without -g, and with/without -emit-module).